api/buckets/ipnsmanager.go

```diff
@@ -0,0 +1,102 @@
+package buckets
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	iface "github.com/ipfs/interface-go-ipfs-core"
+	opt "github.com/ipfs/interface-go-ipfs-core/options"
+	"github.com/ipfs/interface-go-ipfs-core/path"
+	c "github.com/textileio/textile/collections"
+)
+
+// IPNSManager handles bucket name publishing to IPNS.
+type IPNSManager struct {
+	keys     *c.IPNSKeys
+	api      iface.NameAPI
+	lock     sync.Mutex
+	locks    map[string]chan struct{}
+	ctxsLock sync.Mutex
+	ctxs     map[string]context.CancelFunc
+}
+
+// NewIPNSManager returns a new IPNS manager.
+func NewIPNSManager(keys *c.IPNSKeys, api iface.NameAPI) *IPNSManager {
+	return &IPNSManager{
+		keys:  keys,
+		api:   api,
+		ctxs:  make(map[string]context.CancelFunc),
+		locks: make(map[string]chan struct{}),
+	}
+}
+
+// Cancel all pending publishes.
+func (m *IPNSManager) Cancel() {
+	m.lock.Unlock()
+	defer m.lock.Unlock()
+	for _, cancel := range m.ctxs {
+		cancel()
+	}
+}
+
+func (m *IPNSManager) publish(pth path.Path, keyID string) {
+	ptl := m.getSemaphore(keyID)
+	select {
+	case ptl <- struct{}{}:
+		pctx, cancel := context.WithCancel(context.Background())
+		m.ctxsLock.Lock()
+		m.ctxs[keyID] = cancel
+		m.ctxsLock.Unlock()
+		err := m.publishUnsafe(pctx, pth, keyID)
+		if err != nil {
+			if !errors.Is(err, context.Canceled) {
+				// Logging as a warning because this often fails with "context deadline exceeded",
+				// even if the entry can be found on the network (not fully saturated).
+				// The publish deadline seems to be fixed at one minute. ¯\_(ツ)_/¯
+				log.Warnf("error publishing path %s: %v", pth, err)
+			} else {
+				log.Debugf("publishing path %s was cancelled: %v", pth, err)
+			}
+		}
+		cancel()
+		m.ctxsLock.Lock()
+		delete(m.ctxs, keyID)
+		m.ctxsLock.Unlock()
+		<-ptl
+	default:
+		m.ctxsLock.Lock()
+		cancel, ok := m.ctxs[keyID]
+		if !ok {
+			return
```

Comment:
I think I'd structure this method differently.
But a quick solution for now could be:
- Poll `m.ctx[keyID]` a bit with a really small sleep to solve the situation above. So, never return on `!ok`. `!ok`== sleep&retry.
- Call `cancel()` as is in 74
- Avoid de recursive 75, and switch this method to a `for` loop. If we loop more than X times, abort mission since something bad is happening. At least we don't have a StackOverflow panic with the recursive call.

Another idea is after canceling doing `ptl <-struct{}{}`, so the next line knows that the cancelation finished, so can `<-ptl` and retry as is.

We can create an issue to find a more elegant solution later... throwing ideas now to unblock if this in hotpath.

Reply:
Nice, good eye! It's a tricky pattern... like a FILO stack the should always be size one. 

I added a loop to the select statement, negating the recursion, and cleaned up the missed unlock. The default case is allowed to come up empty handed in the _cancel func map_ only `maxCancelPublishTries` times. However, it will run a bunch of times while the publish context is actually cancelling (as you mentioned above, it takes some time... about 2 millis on my machine). I did some testing with consecutive `go pushlish()` calls, and pushing large directories. I've only seen `try` bump to 1. So, I have arbitrarily set `maxCancelPublishTries` to 10.